### ISSUE TITLE
Port twisted.protocols.amp to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6833]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)| @habnabit reported|
|-|-|
|Trac ID|trac#6833|
|Type|enhancement|
|Created|2013-11-14 06:18:13Z|
|Branch|https://github.com/twisted/twisted/tree/amp-python3-6833-4|

`twisted.protocols.amp` is a commonly-used module, and it should run on Python 3.

Attachments:

* [amp-py3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f36/f365960d6be332d3757ce1407b0cb363683da481/5fb26c5b0c9c2bba0cb280254b8a25ff52ba8e54.patch) (5428 bytes) - added by thijs on 2015-11-18 01:17:50Z - 
* [amp-python3-6822-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f36/f365960d6be332d3757ce1407b0cb363683da481/80234ac1a4287b7a7df68f2857cc19d0afc0b0fb.patch) (75942 bytes) - added by allenap on 2015-12-02 15:09:00Z - 
* [amp-python3-6822-3.1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f36/f365960d6be332d3757ce1407b0cb363683da481/a031801b295d16f129f2fb26fb5f5787a08db57b.patch) (90641 bytes) - added by allenap on 2015-12-10 17:41:43Z - 
* [amp-python3-6833-4.add-docstrings.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f36/f365960d6be332d3757ce1407b0cb363683da481/d208ec6ec278704287823c8f40af9f9a623c4284.patch) (1296 bytes) - added by allenap on 2015-12-15 10:56:34Z - Incremental patch to add missing test docstrings to branches/amp-python3-6833-4.

<details><summary>Searchable metadata</summary>

```
trac-id__6833 6833
type__enhancement enhancement
reporter__habnabit habnabit
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_amp_python3_6833_4 branches/amp-python3-6833-4
branch_author__hawkowl__adiroiban hawkowl, adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1384409893000000 1384409893000000
changetime__1450181868838927 1450181868838927
version__None None
owner__allenap allenap
cc__ftsamis@... cc__thijs cc__gavin@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_1"></a>|@habnabit commented|
|-|-|

(In [40739]) Branching to 'amp-python3-6833'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra removed owner|
|-|-|

Attached a patch that fixes the most obvious python 3 errors. It now dies with this error:

```
Traceback (most recent call last):
  File "/Users/thijs/projects/twisted/twisted/trial/runner.py", line 784, in loadByName
    return self.suiteFactory([self.findByName(name, recurse=recurse)])
  File "/Users/thijs/projects/twisted/twisted/trial/runner.py", line 699, in findByName
    obj = reflect.namedModule(searchName)
  File "/Users/thijs/projects/twisted/twisted/python/reflect.py", line 154, in namedModule
    topLevel = __import__(name)
  File "/Users/thijs/projects/twisted/twisted/test/test_amp.py", line 17, in <module>
    from twisted.protocols import amp
  File "/Users/thijs/projects/twisted/twisted/protocols/amp.py", line 2114, in <module>
    _DescriptorExchanger):
  File "/Users/thijs/projects/twisted/twisted/protocols/amp.py", line 2406, in BinaryBoxProtocol
    StartTLS.responder(_defaultStartTLSResponder)
  File "/Users/thijs/projects/twisted/twisted/protocols/amp.py", line 1813, in responder
    CommandLocator._currentClassCommands.append((cls, methodfunc))
builtins.AttributeError: type object 'CommandLocator' has no attribute '_currentClassCommands'
```

Maybe this patch can be send to buildbot so we can start figuring out what's needed to port this.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_4"></a>|@habnabit commented|
|-|-|

That would be because of differing metaclass semantics, which is one of the reasons I gave up when I was working on this before. Since there's interest, here are the two major problems:

1. This problem: metaclasses work differently. At the time I was working on this, six only had `with_metaclass`, which was insufficient. I think the newer `@add_metaclass` will do the right thing, though.
1. AMP takes advantage of that python native strings are bytes. Since python 3 native strings are not bytes, there's a _lot_ of code that breaks. With the size of AMP's implementation and tests, it's tedious to examine each string literal to determine "okay, is this supposed to be bytes or a native string?" I don't know if semantics were ever defined for non-ASCII identifiers in AMP, but it might be sufficient to force utf-8 for all keys.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

(In [46279]) Branching to amp-py3-6833.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

(In [46280]) Branching to amp-python3-6833-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to @thijstriemstra|
|-|-|

Patch applied and sent to buildbot.

Looking forward for the final patch.

Don't forget to add a news file fragment and described here http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|allenap removed owner|
|-|-|

This is a port of `twisted.protocols.amp` to Python 3. The tests pass on Python 2.7, 3.4 and 3.5. My branch is on GitHub, and a diff can be seen here:

  https://github.com/twisted/twisted/compare/trunk...allenap:amp-python3-6833-3

I did see thijs's work on this before starting; his/her contribution was helpful. Thank you!

It began as fairly strict with regards to byte strings: if something MUST be a byte string or COULD be a byte string in Python 2, it MUST be a byte string in Python 3.

However, I relaxed that rule in several instances for Python 3, to create a more natural API and, in some instances, out of necessity:

* `AmpBox.__init__()` accepts keyword arguments. In Python 3 these MUST be
  Unicode/native strings whereas in Python 2 they could be either byte
  strings or Unicode strings. In Python 3, therefore, keys are coerced
  to (ASCII) byte strings.

  This poses a problem: `AmpBox.__init__()` has no command information so
  there's no way to make use of `_wireNameToPythonIdentifier()`. Coercing
  is all we can do.

  However, this coercion is entirely for the benefit of tests. The
  implementation code does not use keyword arguments when creating
  instances of `AmpBox` and its derivatives. Eliminating the use of
  keyword arguments when creating `AmpBox` would address the need for
  coercion. There may be non-Twisted code around that creates `AmpBox`
  directly, so a compromise might be to forbid keyword arguments under
  Python 3 only.

* `BoxDispatcher._errorReceived()` decodes error descriptions as UTF-8
  text using the //replace// errors strategy. Python 3 does not permit
  byte strings as exception descriptions.

  Would the //surrogateescape// errors strategy be better, so that
  decoding is loss-less?

* `BoxDispatcher._commandReceived()` > `formatError()` encodes error
  descriptions as UTF-8 text using the //replace// errors strategy.

  Would the //surrogateescape// errors strategy be better, so that
  encoding is loss-less?

At the same time as doing this port I've also been porting some
not-in-Twisted code that uses `t.p.amp`. This highlighted some mistakes I
made early on, and has helped this port feel "right".

I had to make many changes to the tests, mostly to add binary prefixes
to strings. However, I have tested the ported `t.p.amp` module with the
non-ported test module, and it worked (in Python 2.7 only, naturally).

There's one small thing that I spotted that I'm not sure how to fix, or
even if we care to fix it:

* `Path` (`Argument` subclass): A bytes-mode `FilePath` on one side will
  always be deserialised as a text-mode `FilePath`.

Of course, there may be others.

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

Update branch

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban commented|
|-|-|

I have never used AMP so I am just doing a high level review here.

Tests are green ... but the test code itself was also updated.. so I am not sure if actual application using AMP will still work after we merge these changes.

-------

Do we still need the ``__metaclass__ = type`` declaration?

Do we want/need byteString as a public API?

```AmpBox.__init___``` special Python3 handling needs a comment.

I know that you added the comments as a Trac comments but we need this info as close as possible to the source and future developers will not browser all past Trac conversation to get info about a certain change.

Also, please check Twisted naming conventions non_byte_names vs nonByteNames

Does the new AMP code only accept bytes or still works with 'str' on Py3 ?
Please also update the docstrins to make it clear that we have 'bytes'.

While on Py2 str is just bytes, on Py3 str closer to unicode than bytes. 

If a value should be unicode, please keep the `u"ignored"` marker as it will make sure it continue to be Unicode in Python2.


--------

For BoxDispatcher._nextTag can we have some call which works on all systems... example explicit hexify ?
From what I can see, there is no interpolation there, just a conversion to hex... which as a hack it was implemented using interpolation.

I would like to see as few _PY3 branches as possible.

--------

In `dispatchCommand` the interpolation is just used as a concatenation. Can we replace the bytesString call with a bytes concatenation?

-----


```
# Apply __metaclass__ to CommandLocator.
```

Is just stating the obvious, can we have a comment explaining why this is needed ? :)


------


For `_wireNameToPythonIdentifier`, why do we need the new assertion?

Can we just update the docstring to inform that bytes are required ?

This looks like a private method so we are in full control of the way in which it is called.

-------


For Argument.fromBox / toBox we need to update the docstrings to explicilty say bytes....
and this is a general comments for all the current API.

But is it ok to force bytes ? What happen with existing apps which pass Unicode ?

----------

I see a lot of new `assert` statments. Why do we need them?

I prefer to have better docstrings that those assertions.

--------

Here is the code coverage

https://codecov.io/github/twisted/twisted/commit/a78c07355c15a7e7cb24b3beea7f0c32846800a1

Please note that the port should have at least 100% coverage ... so if the new code raises new exceptions (including assertions) please write tests for them.

-----


Here are some style guide errors for our builder:

```

(view as text)
************* Module twisted.protocols.amp
W9201:293,0:byteString: The opening/closing of docstring should be on a line by themselves
W9202:293,0:byteString: Missing epytext markup @param for argument "encoding"
W9202:293,0:byteString: Missing epytext markup @param for argument "string"
C0103:624,12:AmpBox.__init__: Invalid name "non_byte_names" for type variable (should match ([a-z_][a-zA-Z0-9]*)$)
C0103:625,16:AmpBox.__init__: Invalid name "non_byte_name" for type variable (should match ([a-z_][a-zA-Z0-9]*)$)
C0103:626,16:AmpBox.__init__: Invalid name "byte_name" for type variable (should match ([a-z_][a-zA-Z0-9]*)$)
W9501:823,20:BoxDispatcher._nextTag: String formatting operations should always use a tuple for non-mapping values
************* Module twisted.test.test_amp
C0103:3068,8:DescriptorTests.test_roundTrip: Invalid name "name_as_bytes" for type variable (should match ([a-z_][a-zA-Z0-9]*)$)


```

----------

I am leaving this for review as I hope that this patch will get a second pair of eyes... but please consider my comments and update your branch.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to allenap|
|-|-|

Please check my comments and submit a new patch.

And I hope that we can find someone else to do the final review.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|allenap set owner to @adiroiban|
|-|-|

My comments below refer to the (quoted) review of _amp-python3-6822-3.patch_ and my updates which are in _amp-python3-6822-3.1.patch_.

>
> I have never used AMP so I am just doing a high level review here.
>
> Tests are green ... but the test code itself was also updated.. so I
> am not sure if actual application using AMP will still work after we
> merge these changes.
>
> -------
>
> Do we still need the `__metaclass__ = type` declaration?

For Python 2, yes, for the classes that are not defined with an explicit
metaclass.

>
> Do we want/need byteString as a public API?

No. I realised it was used in two places and neither greatly benefitted
from it. I have removed it.

>
> `AmpBox.__init___` special Python3 handling needs a comment.
>
> I know that you added the comments as a Trac comments but we need this
> info as close as possible to the source and future developers will not
> browser all past Trac conversation to get info about a certain change.

Good point. Done.

>
> Also, please check Twisted naming conventions non_byte_names vs
> nonByteNames

Done.

>
> Does the new AMP code only accept bytes or still works with 'str' on
> Py3 ? Please also update the docstrins to make it clear that we have
> 'bytes'.

In general it accepts native strings in Python 3, but coerces them to
byte strings where necessary. In many places there's no need to coerce
because `_wireNameToPythonIdentifier` is used to derive a key. When
updating the docs I've drawn attention to places where this happens.

>
> While on Py2 str is just bytes, on Py3 str closer to unicode than
> bytes.
>
> If a value should be unicode, please keep the `u"ignored"` marker as
> it will make sure it continue to be Unicode in Python2.

I think there was just the one place where I had inadvertently removed
the `u` prefix. Fixed, but let me know if you found others.

>
>
> --------
>
> For BoxDispatcher._nextTag can we have some call which works on all
> systems... example explicit hexify ?
> From what I can see, there is no interpolation there, just a
> conversion to hex... which as a hack it was implemented using
> interpolation.
>
> I would like to see as few _PY3 branches as possible.

`hexlify` operates on byte strings. `hex(int)` returns a native string so
this would need to be encoded in Python 3 too (it also returns a leading
"0x" which would need to be removed). We could use a Unicode string in
all Python versions:

```
    def _nextTag(self):
        self._counter += 1
        return (u'%x' % self._counter).encode("ascii")
```

which would avoid the conditional on `_PY3`. What do you think?

>
> --------
>
> In `dispatchCommand` the interpolation is just used as a
> concatenation. Can we replace the bytesString call with a bytes
> concatenation?

I've eliminated this. The description passed to `RemoteAmpError` should
always be a native string, and it is elsewhere.

>
> -----
>
>
> ```
> # Apply __metaclass__ to CommandLocator.
> ```
>
> Is just stating the obvious, can we have a comment explaining why this
> is needed ? :)

Done.

>
>
> ------
>
>
> For `_wireNameToPythonIdentifier`, why do we need the new assertion?
>
> Can we just update the docstring to inform that bytes are required ?
>
> This looks like a private method so we are in full control of the way
> in which it is called.

It was a belt-n-braces thing. It helped while I was working but can be
removed now.

>
> -------
>
>
> For Argument.fromBox / toBox we need to update the docstrings to
> explicilty say bytes.... and this is a general comments for all the
> current API.

Done.

>
> But is it ok to force bytes ? What happen with existing apps which pass
> Unicode ?

I added most/all of the assertions to help with the port. Most of them
I've now removed.

`Unicode.toString()`

  In Python 2 folk could be taking advantage of implicit promotion from
  str to unicode. In Python 3 this will break if the argument is a byte
  string. The assertion is no longer necessary, and could break some
  valid (though dubious) scenarios in Python 2.

`AmpList.__init__()`

  This checks that all sub-arguments are defined with byte string names.
  This helps catch mistakes early on; errors resulting from a mistake
  here can be hard to diagnose. I've improved the fail message.

>
> ----------
>
> I see a lot of new `assert` statments. Why do we need them?

I added most/all of the assertions to help with the port, explanations
for those remaining are above.

>
> I prefer to have better docstrings that those assertions.

I agree with the docstrings.

>
> --------
>
> Here is the code coverage
>
> https://codecov.io/github/twisted/twisted/commit/a78c07355c15a7e7cb24b3beea7f0c32846800a1
>
> Please note that the port should have at least 100% coverage ... so if
> the new code raises new exceptions (including assertions) please write
> tests for them.

Done, I think. Most significantly I added new tests for the checks I previously added to `Command.__metaclass__`.

>
> -----
>
>
> Here are some style guide errors for our builder:
>
> ```
> ************* Module twisted.protocols.amp
> W9201:293,0:byteString: The opening/closing of docstring should be on a line by themselves
> W9202:293,0:byteString: Missing epytext markup @param for argument "encoding"
> W9202:293,0:byteString: Missing epytext markup @param for argument "string"
> C0103:624,12:AmpBox.__init__: Invalid name "non_byte_names" for type variable (should match ([a-z_][a-zA-Z0-9]*)$)
> C0103:625,16:AmpBox.__init__: Invalid name "non_byte_name" for type variable (should match ([a-z_][a-zA-Z0-9]*)$)
> C0103:626,16:AmpBox.__init__: Invalid name "byte_name" for type variable (should match ([a-z_][a-zA-Z0-9]*)$)
> W9501:823,20:BoxDispatcher._nextTag: String formatting operations should always use a tuple for non-mapping values
> ************* Module twisted.test.test_amp
> C0103:3068,8:DescriptorTests.test_roundTrip: Invalid name "name_as_bytes" for type variable (should match ([a-z_][a-zA-Z0-9]*)$)
> ```

I think I've addressed all of those.

>
> ----------
>
> I am leaving this for review as I hope that this patch will get a
> second pair of eyes... but please consider my comments and update your
> branch.

Thank you!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

(In [46462]) Branching to amp-python3-6833-4.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_15"></a>|@hawkowl commented|
|-|-|

(In [46466]) applying patch from allenap, refs [#6833](https://github.com/twisted/twisted/issues/6833)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl set owner to allenap|
|-|-|

Hi allenap!

Thanks for the patch! The coverage is looking very nice, so I'm confident that it'll work great when merged :)

Unfortunately twisted.protocols.amp relies on twisted.python._tzhelper, which is not a ported module. Since it's small enough, it should be able to be ported as a part of this ticket with no worries.

Could you port it (basing your patch off this branch, as I fixed up a future import or two) and put it up for review again?

Again, many thanks! This is an essential module that a lot of people use, and it being ported is a good first step for other modules which depend on it.

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|allenap removed owner|
|-|-|

I'm happy to do that, but I think t.p._tzhelper has been ported already. It (and test_tzhelper) are in dist3.py, and the tests pass in Python 3.4 and 3.5.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_18"></a>|@adiroiban commented|
|-|-|

Thanks. I can confirm that tzhelper is in dist3.pyy 

Waiting for Amber to come online just to double check.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_19"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Huh, seems it is. Doesn't have __future__ imports, but if it's in dist3, it probably works.

Adi, go ahead and land it :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_20"></a>|@adiroiban set owner to allenap|
|-|-|

Thanks, will merge.

Thanks Gavin for your contribution!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_21"></a>|@adiroiban commented|
|-|-|

Waiting for test results from latest commit... and there is a failing documentation builder.

**Comment by twisted-trac** [comment]
|<img alt="allenap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|allenap set owner to @adiroiban|
|-|-|

Incremental patch for docstrings, as requested in #twisted-dev. Plus I have only just realised that I had named my patches with the wrong ticket number. Oops, and sorry.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_23"></a>|@adiroiban set owner to allenap|
|-|-|

Thanks. Latest patch looks good.

I will merge.

Many thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_24"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46516]) Merge amp-python3-6833-4: Port twisted.protocols.amp to Python 3.

Author: allenap
Reviewer: adiroiban, hawkowl
Fixes: [#6833](https://github.com/twisted/twisted/issues/6833)
